include/vkwrap/vkwrap/image.h

```diff
@@ -0,0 +1,316 @@
+
+}; // class Image
+
+class ImageBuilder
+{
+
+  public:
+    using Queues = std::vector<Queue>;
+
+    // clang-format off
+    PATCHABLE_DEFINE_STRUCT( 
+        ImagePartialInfo,
+        ( std::optional<vk::ImageType>,           image_type   ),
+        ( std::optional<vk::Format>,              format       ),
+        ( std::optional<vk::Extent3D>,            extent       ),
+        ( std::optional<uint32_t>,                array_layers ),
+        ( std::optional<vk::SampleCountFlagBits>, samples      ),
+        ( std::optional<vk::ImageTiling>,         tiling       ),
+        ( std::optional<vk::ImageUsageFlags>,     usage        ),
+        ( std::optional<Queues>,                  queues       )
+    );
+    // clang-format on
+
+    using Setter = std::function<void( ImagePartialInfo& )>;
+
+  private:
+    // clang-format off
+    Setter m_setter = []( auto& ){};
+    // clang-format on
+
+    ImagePartialInfo m_partial;
+
+    ImagePartialInfo makePartialInfo() const
+    {
+        ImagePartialInfo partial{};
+
+        presetter( partial );
+        m_setter( partial );
+        partial.patchWith( m_partial );
+
+        return partial;
+    } // makePartialInfo
+
+    static constexpr vk::ImageCreateInfo k_initial_create_info = {
+        // We don't use these specific fields.
+        .pNext = {},
+        .flags = {},
+
+        // We don't use compressed images.
+        .mipLevels = 1,
+
+        // Initial layout is unknown.
+        .initialLayout = vk::ImageLayout::eUndefined };
+
+  public:
+    // clang-format off
+    static inline Setter presetter = []( auto& ){};
+    // clang-format on
+
+    ImageBuilder() = default;
+
+    ImageBuilder& withSetter( Setter setter ) &
+    {
+        m_setter = setter;
+        return *this;
+    } // withSetter
+
+    ImageBuilder& withImageType( vk::ImageType type ) &
+    {
+        m_partial.image_type = type;
+        return *this;
+    } // withImageType
+
+    ImageBuilder& withFormat( vk::Format format ) &
+    {
+        m_partial.format = format;
+        return *this;
+    } // withFormat
+
+    ImageBuilder& withExtent( vk::Extent3D extent ) &
+    {
+        m_partial.extent = extent;
+        return *this;
+    } // withExtent
+
+    ImageBuilder& withArrayLayers( uint32_t array_layers )
+    {
+        assert( array_layers >= 1 );
+
+        m_partial.array_layers = array_layers;
+        return *this;
+    } // withArrayLayers
+
+    ImageBuilder& withSampleCount( vk::SampleCountFlagBits samples ) &
+    {
+        m_partial.samples = samples;
+        return *this;
+    } // withSampleCount
+
+    ImageBuilder& withTiling( vk::ImageTiling tiling ) &
+    {
+        m_partial.tiling = tiling;
+        return *this;
+    } // withTiling
+
+    ImageBuilder& withUsage( vk::ImageUsageFlags usage ) &
+    {
+        m_partial.usage = usage;
+        return *this;
+    } // withUsage
+
+    ImageBuilder& withQueues( ranges::range auto&& queues ) &
```

Comment:
Better add a constraint on the range_value_t to be the same as vkwrap::Queue

Reply:
As in this function `m_partial.queues = ranges::to_vector( queues );` and `m_partial.queues` is `std::vector<vkwrap::Queue>`, I don't think it would compile if `ranges_value_t != vkwrap::Queue`.
But ok, I've added concept.